### DISCUSSION TITLE
More robust substitution system

### DISCUSSION THREAD

**Discussion by penguincrest** [karlomikus/bar-assistant#300] [Ideas]
I'm loving Bar Assistant and there have so far only been one big thing that could be changed.  Right now it appears substitutions are only done based on the parent ingredient category.  It might be helpful to add "flavor" tags to ingredients.  For instance, some ingredients may share the same category but be completely different flavor profiles.  Some system of allowing ingredients to be in a category but have additional tags to narrow down usefulness.  Such as under Liqueurs, you could have nut, fruit, or coffee liqueurs.  Maybe tags in a similar way that cocktails have tags.

That could also allow users to get more specific in their shelf list, like we could add specific bottles instead of a generic "rum" or "whisky" and just add tags to the bottles.  That also opens the door to add specific bottles for whisky collections.

**Comment by zhdenny** [comment]
In any given drink recipe, I can provide substitutes for any ingredient. This allows me to be as detailed/robust as I want to be.
![Screenshot_20240709-092542](https://github.com/karlomikus/bar-assistant/assets/26662131/fcd90d86-0af2-4bfa-8c0b-ead9b792a2af)

**Comment by zhdenny** [comment]
In addition, I do in fact add specific bottles. And I can already declare them as part of a larger group.

In this case "Smith & Cross" Rum is a variety of "Rum - Jamaican Lightly Aged" in the "Rum - Jamaican" Category.

My point here is I think this app has the tools necessary to make pretty robust categorization.
I heavily modified my database to do this. The default stuff leaves much to be desired.
![Screenshot_20240709-093310](https://github.com/karlomikus/bar-assistant/assets/26662131/629f3115-a789-4372-9feb-38f6bd27a40b)

**Comment by zhdenny** [comment]
Spirits and beverages categorization can be a complicated process that has no one-size-fits-all answer.

For rum I use the category solution outlined in the popular book Smuggler's Cove by Martin Cate.

**Reply by zeekaran** [reply]
Just to clarify, is this how you do it?

Categories:

- Rum
- Rum - Jamaican
- Rum - Barbados (Barbadian?)
- Rum - Haitian

(I expect you do not have a "Spirits" category anymore?)

Ingredients:

- Rum - Black Blended Overproof (category: Rum)
- Plantation OFTD (parent: Rum - Black Blended Overproof) (category: Rum)
- Hamilton 151 (parent: Rum - Black Blended Overproof) (category: Rum)
- Lemonhart 151 (parent: Rum - Black Blended Overproof) (category: Rum)
- Rum - Blended Lightly Aged
- Rum - Blended Aged

**Reply by zeekaran** [reply]
And how might you add this cocktail?

## Bombo

**Recommended spirits:**
Plantation OFTD Rum
Pusser's Gunpowder Proof British Navy Rum
Navy Rum Blend

- 1 bar spoon cinnamon syrup
- 2 ounces Navy-style rum

Garnish: Whole nutmeg, grated

**Reply by zeekaran** [reply]
Last question (for now), do you add every single spirit called for by a recipe as a new ingredient, and just add the appropriate parent? Because so far all two recipes I've added I just put the parent as the ingredient, but add the specific recommended bottle as a note on the ingredient.

I expect you do so that you can use filter and search features on specific ingredients, where my way means I can't filter by drinks with Plantation OFTD because it's merely a note.

**Comment by zhdenny** [comment]
You are more than welcome to import my database. I upload it to dropbox daily. Full import instructions are provided in my github.

https://github.com/zhdenny/bar_assistant_database

But I have a series of broad Rum Ingredient Categories......
![image](https://github.com/user-attachments/assets/907ce8da-3b56-4631-9d45-75e60f3ea978)


Then I create individual ingredients for sub categories....For example, this is an ingredient. I provide examples of specific bottles and a full explanation it the sub category.
![image](https://github.com/user-attachments/assets/7a1d336e-f929-491d-a75e-6cecc27b7d6d)


Then if I call out specific bottles regularly throughout multiple drinks, I make an ingredient for the specific bottle and make sure to call it out as a variety of the sub category ingredient. For example:
![image](https://github.com/user-attachments/assets/32b98b30-8546-4f10-b796-a77a04efea8f)

To be perfectly clear, I do this level of detail primarily for RUM. Rum is extremely diverse compared to other primary spirits [Whisk(e)y, Gin, Vodka, Brandy, etc]. Rum really does require finer detail so I branch out categories and sub-categories.

If a drink really does need and call for a specific bottle and it is something I don't use often, I will call on the sub-category ingredient and put a note for the specific bottle asked for.

I spent quite a lot of time and effort on my database......There are a handful of users who simply import my database into their BarAssistant/Saltrim instance.

**Reply by zeekaran** [reply]
Thanks, I just deleted my whole setup to use yours instead. You have a typo on "Hot Suace" by the way.

**Reply by zhdenny** [reply]
> Thanks, I just deleted my whole setup to use yours instead. You have a typo on "Hot Suace" by the way.

Fixed and Updated. Thanks for reporting.

**Comment by wildegnux** [comment]
I've imported your database @zhdenny but I still find the substitution system lacking... 
I have added `Smith & Cross`, which is a variety of `Rum - Jamaican Lightly Aged`, to my shelf. So if a recipe calls for the ingredient `Rum - Jamaican Lightly Aged` then I'd like the ingredients list for the recipe to indicate that I do in fact have that. Same with the `Cocktails I can make` feature. I have smith & cross on my shelf so show me the cocktails I can make, including ones that require `Rum - Jamaican Lightly Aged`
![image](https://github.com/user-attachments/assets/5d63c1c8-3c90-413f-9995-ecb666c7fe22)
![image](https://github.com/user-attachments/assets/3986d06a-a4c8-449f-ad4f-e5762c979ba4)

**Reply by zhdenny** [reply]
Personally, when I have Smith & Cross, I add both items onto my shelf

- Rum - Jamaican Lightly Aged
- Smith & Cross Rum

But just generally speaking, I try hard not to add specific bottles as ingredients in my database. Ultimately, maybe I should just remove Smith & Cross from my ingredients list entirely and use the general rum group ingredient instead (Rum - Jamaican Lightly Aged).

I can do that if you want.
